sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ColumnFamilySchemaUtils.scala

```diff
@@ -17,58 +17,68 @@
 package org.apache.spark.sql.execution.streaming
 
 import org.apache.spark.sql.Encoder
-import org.apache.spark.sql.execution.streaming.TransformWithStateKeyValueRowSchema.{COMPOSITE_KEY_ROW_SCHEMA, KEY_ROW_SCHEMA, VALUE_ROW_SCHEMA, VALUE_ROW_SCHEMA_WITH_TTL}
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+import org.apache.spark.sql.execution.streaming.TransformWithStateKeyValueRowSchema._
+import org.apache.spark.sql.execution.streaming.TransformWithStateKeyValueRowSchema.{COMPOSITE_KEY_ROW_SCHEMA, KEY_ROW_SCHEMA}
 import org.apache.spark.sql.execution.streaming.state.{ColumnFamilySchema, ColumnFamilySchemaV1, NoPrefixKeyStateEncoderSpec, PrefixKeyScanStateEncoderSpec}
 
 trait ColumnFamilySchemaUtils {
-  def getValueStateSchema[T](stateName: String, hasTtl: Boolean): ColumnFamilySchema
+  def getValueStateSchema[T](
+      stateName: String,
+      keyEncoder: ExpressionEncoder[Any],
+      valEncoder: Encoder[T],
+      hasTtl: Boolean): ColumnFamilySchema
 
-  def getListStateSchema[T](stateName: String, hasTtl: Boolean): ColumnFamilySchema
+  def getListStateSchema[T](
+      stateName: String,
+      keyEncoder: ExpressionEncoder[Any],
+      valEncoder: Encoder[T],
+      hasTtl: Boolean): ColumnFamilySchema
 
   def getMapStateSchema[K, V](
       stateName: String,
+      keyEncoder: ExpressionEncoder[Any],
       userKeyEnc: Encoder[K],
+      valEncoder: Encoder[V],
       hasTtl: Boolean): ColumnFamilySchema
 }
-
 object ColumnFamilySchemaUtilsV1 extends ColumnFamilySchemaUtils {
 
-  def getValueStateSchema[T](stateName: String, hasTtl: Boolean): ColumnFamilySchemaV1 = {
-    ColumnFamilySchemaV1(
+  def getValueStateSchema[T](
+      stateName: String,
+      keyEncoder: ExpressionEncoder[Any],
+      valEncoder: Encoder[T],
+      hasTtl: Boolean): ColumnFamilySchemaV1 = {
+    new ColumnFamilySchemaV1(
       stateName,
-      KEY_ROW_SCHEMA,
-      if (hasTtl) {
-        VALUE_ROW_SCHEMA_WITH_TTL
-      } else {
-        VALUE_ROW_SCHEMA
-      },
+      getKeySchema(keyEncoder.schema),
+      getValueSchemaWithTTL(valEncoder.schema, hasTtl),
       NoPrefixKeyStateEncoderSpec(KEY_ROW_SCHEMA))
   }
 
-  def getListStateSchema[T](stateName: String, hasTtl: Boolean): ColumnFamilySchemaV1 = {
-    ColumnFamilySchemaV1(
+  def getListStateSchema[T](
+      stateName: String,
+      keyEncoder: ExpressionEncoder[Any],
+      valEncoder: Encoder[T],
+      hasTtl: Boolean): ColumnFamilySchemaV1 = {
+    new ColumnFamilySchemaV1(
       stateName,
-      KEY_ROW_SCHEMA,
-      if (hasTtl) {
-        VALUE_ROW_SCHEMA_WITH_TTL
-      } else {
-        VALUE_ROW_SCHEMA
-      },
+      getKeySchema(keyEncoder.schema),
+      getValueSchemaWithTTL(valEncoder.schema, hasTtl),
       NoPrefixKeyStateEncoderSpec(KEY_ROW_SCHEMA))
```

Comment:
Do we need to change the schema passed to the encoder spec too ?

Reply:
I think the encoder takes in the BinaryType schema: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImpl.scala#L48
I left a comment [here](https://github.com/apache/spark/blob/b57575ba951d690b97ada9056aba6e5a9e9c2968/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala#L29)

Reply:
The state store eventually will need to read/write byte arrays. but we can deal with native `StructType` schemas passed at this layer. Discussed offline with @jingz-db